### PR TITLE
[libgig] New Port

### PR DESCRIPTION
**Pull request by Rossmaxx**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [ ] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [ ] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

Hello. I am a volunteer from the LMMS project and we use libgig to build an audio plugin. We are trying to add complete MSVC support for LMMS. As a requirement, @Veratil and @PhysSong have written this port (I'm forwarding it upstream on their behalf). Hoping this port gets accepted here.

### CODE REVIEW

**Comment by Rossmaxx** [comment]
@microsoft-github-policy-service agree

**Comment by Rossmaxx** [comment]
I don't understand why CI is failing. The port works in my pc fine. Could someone help. Also, I'm new to submitting ports. So if I missed something, please correct me.

**Comment by dg0yt** [comment]
To fix osx, require at least C++11 in CMakeLists.txt.
uwp seems unsupported, and this can be declared in vcpkg.json.
Failure logs can be downloaded from the CI build's published artifacts. The download button is in a menu behind three dots on the right side when hovering an artifact.
https://dev.azure.com/vcpkg/public/_build/results?buildId=93110&view=artifacts&pathAsName=false&type=publishedArtifacts

**Comment by Rossmaxx** [comment]
> require at least C++11 in CMakeLists.txt.

Should i forward it upstream or use a patch for it.

> uwp seems unsupported, and this can be declared in vcpkg.json.

How do i do it? Any example files which i could look at. 

Also, what about 32 bit windows? 

Thanks for replying. Will do it all and push it.

**Comment by PhysSong** [comment]
@Rossmaxx If you don't mind, can I continue fixing errors on a new PR?
Note to maintainers: I'm the person who've added a few patches to @Veratil's original version, and the version in this PR is the same as my latest work.

**Comment by Rossmaxx** [comment]
> If you don't mind, can I continue fixing errors on a new PR?

Sure. Just inform me once you open that one. I'll close this pr then. But if possible, can you pr against my branch to reduce duplicates.

**Comment by Rossmaxx** [comment]
> uwp seems unsupported, and this can be declared in vcpkg.json.

done

**Comment by Rossmaxx** [comment]
Weird. 32 bit windows still fails.

**Comment by Rossmaxx** [comment]
Patched 32 bit builds failing too. It should be ready now.

**Comment by Rossmaxx** [comment]
Apparently, it's a SHA checksum error which occurs only on 32 but windows. I'm confused now.

**Comment by dg0yt** [comment]
That's a check for the proper versions entry. Try
~~~
.\vcpkg x-add-version --overwrite-version libgig
git add -u versions
git commit -m "Update versions"
git push

**Comment by LilyWangLL** [comment]
```
    find_package(libgig CONFIG REQUIRED)
    target_link_libraries(main PRIVATE libgig::libgig libgig::libakai)
```
Usage test failed with the following error:
```
1> [CMake]   Target "PRreview" links to:
1> [CMake] 
1> [CMake]     libgig::libakai
1> [CMake] 
1> [CMake]   but the target was not found.  Possible reasons include:
1> [CMake] 
1> [CMake]     * There is a typo in the target name.
1> [CMake]     * A find_package call is missing for an IMPORTED target.
1> [CMake]     * An ALIAS target is missing.
```
Could you please also resolve this?

**Comment by LilyWangLL** [comment]
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

**Comment by LilyWangLL** [comment]
Please also use [vcpkg_copy_tools](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/maintainers/functions/vcpkg_copy_tools.md) handle the following tools when building feature `extra-tools`.
```
add_tool(dlsdump)
add_tool(gigdump)
add_tool(gigmerge)
add_tool(korg2gig)
add_tool(korgdump)
add_tool(rifftree)
add_tool(sf2dump)
```

**Comment by Rossmaxx** [comment]
Actually, we don't need libakai for now, so I might consider removing it for now (if no one objects it)

**Comment by PhysSong** [comment]
> Usage test failed with the following error:

The library provides the separate `share/libgig/libakai-config.cmake` file for `libgig::libakai`. What's the recommended fix?

**Comment by LilyWangLL** [comment]
> > Please also use vcpkg_copy_tools handle the following tools when building feature extra-tools.
> > --Code--
> 
> Should i add that to the portfile near it's end should i do anything else.

Please add the following codes below `vcpkg_cmake_install()`.
```
if("extra-tools" IN_LIST FEATURES)
    vcpkg_copy_tools(TOOL_NAMES dlsdump gigdump gigmerge korg2gig korgdump rifftree sf2dump
        SEARCH_DIR "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/bin"
        AUTO_CLEAN)
endif()
```

File: ports/libgig/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+        ${FEATURE_OPTIONS}
+        -DLIBGIG_BUILD_SHARED=${LIBGIG_BUILD_SHARED}
+)
+
+vcpkg_cmake_install()
+vcpkg_fixup_pkgconfig()
```
Comment (LilyWangLL) [inline ports/libgig/portfile.cmake:32]:
```suggestion
```
There are no pc files be installed by this port.

**Review by LilyWangLL — CHANGES_REQUESTED** [review]

**Comment by LilyWangLL** [comment]
> > Usage test failed with the following error:
> 
> The library provides the separate `share/libgig/libakai-config.cmake` file for `libgig::libakai`. What's the recommended fix?

You can change the codes `install(TARGETS libakai...` in file **[SOURCE_PATH]/CMakeLists.txt** to the following code.
```
install(TARGETS libakai EXPORT libgig-config
	LIBRARY DESTINATION lib
	ARCHIVE DESTINATION lib
	RUNTIME DESTINATION bin
	PUBLIC_HEADER DESTINATION include/libgig)
install(EXPORT libgig-config NAMESPACE libgig:: DESTINATION share/libgig)
```
BTW, this should be an upstream issue, also need fix by upstream.

**Comment by LilyWangLL** [comment]
If you have new change, please run command `.\vcpkg x-add-version libgig --overwrite-version`, this will update `versions/l-/libgig.json`. The CI failure is about this.

**Comment by Rossmaxx** [comment]
> You can change the codes install(TARGETS libakai... in file [SOURCE_PATH]/CMakeLists.txt to the following code.
-- code --

I patched that in temporarily but when I patched, CI was complaining so reverted the patch. Also addressed the other issues.

**Comment by Rossmaxx** [comment]
How do i resolve the existing libakai issues?

**Comment by LilyWangLL** [comment]
> I patched that in temporarily but when I patched, CI was complaining so reverted the patch. Also addressed the other issues.

The CI failure is about change version, please run command `.\vcpkg x-add-version libgig --overwrite-version`, then commit to the branch.

> If you have new change, please run command `.\vcpkg x-add-version libgig --overwrite-version`, this will update `versions/l-/libgig.json`. The CI failure is about this.

**Comment by LilyWangLL** [comment]
> How do i resolve the existing libakai issues?

Fix as below.
> You can change the codes `install(TARGETS libakai...` in file **[SOURCE_PATH]/CMakeLists.txt** to the following code.
> 
> ```
> install(TARGETS libakai EXPORT libgig-config
> 	LIBRARY DESTINATION lib
> 	ARCHIVE DESTINATION lib
> 	RUNTIME DESTINATION bin
> 	PUBLIC_HEADER DESTINATION include/libgig)
> install(EXPORT libgig-config NAMESPACE libgig:: DESTINATION share/libgig)
> ```
> 
> BTW, this should be an upstream issue, also need fix by upstream.

**Comment by Rossmaxx** [comment]
> I patched that in temporarily but when I patched, CI was complaining.

Still true. 

> The CI failure is about change version

Nope it ain't it. Even after running the version command, still it's failing.

**Comment by Rossmaxx** [comment]
So the change you suggested might be unnecessary, instead there might be a need to either restrict support for windows only or refactor the entire cmake file to enable stuff in a better way.

**Comment by Rossmaxx** [comment]
Found a github clone of the upstream code [here](https://github.com/drbye78/libgig). @LilyWangLL can you suggest the best change needed. Also, how can I reproduce the issue?

**Comment by Rossmaxx** [comment]
Since I've not got any replies from @LilyWangLL for the past 2 weeks, I've decided to revert the attempted fixes at libakai. His suggestions do not work and is failing ci (even after updating version). If Lily (or anyone else) have got any other suggestion, please let me know so that I can complete this PR.

**Comment by Rossmaxx** [comment]
I have decided to shortlist the support to windows alone as libgig already exists in homebrew and also in the package managers for almost all linux distros.

**Comment by LilyWangLL** [comment]
[fix-usage.patch](https://github.com/microsoft/vcpkg/files/12746652/fix-usage.patch)

Sorry for the late reply. Please add this patch to this PR, it will fix the usage.

**Comment by Rossmaxx** [comment]
It seems this patch works now, I now understand why. When I tried to patch, I didn't remove the first line then changed just the `libakai:config` to `libgig:config`.

File: ports/libgig/vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
+    }
+  ],
+  "features": {
+    "extra-tools": {
+      "description": "Build extra tools",
```
Comment (dg0yt) [inline ports/libgig/vcpkg.json:19]: Which tools are "normal", and which are "extra"?
(I don't see any post-processing for tools other than `extra-tools`.)

File: ports/libgig/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    OPTIONS
+        ${FEATURE_OPTIONS}
+        -DLIBGIG_BUILD_SHARED=${LIBGIG_BUILD_SHARED}
+)
+
+vcpkg_cmake_install()
+if("extra-tools" IN_LIST FEATURES)
+    vcpkg_copy_tools(TOOL_NAMES dlsdump gigdump gigmerge korg2gig korgdump rifftree sf2dump
+        SEARCH_DIR "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/bin"
+        AUTO_CLEAN)
```
Comment (dg0yt) [inline ports/libgig/portfile.cmake:35]: We don't need to auto-clean the build dir.

Comment (Rossmaxx) [reply ports/libgig/portfile.cmake:35]: Is it fine to delete those parts?

Comment (Rossmaxx) [reply ports/libgig/portfile.cmake:35]: Is it fine to remove the  `AUTO_CLEAN` at the end?

File: ports/libgig/vcpkg.json (hunk 2 above)
Comment (Rossmaxx) [reply ports/libgig/vcpkg.json:19]: Not clear, what exactly am i supposed to do here?

Comment (dg0yt) [reply ports/libgig/vcpkg.json:19]: "Extra" is not clear. Either explain which tools are not controlled by "extra". Or remove the "extra" letters. Many ports have tools, and every feature is "extra".

File: ports/libgig/portfile.cmake (hunk 3 above)
Comment (dg0yt) [reply ports/libgig/portfile.cmake:35]: Yes, it is okay here. (That's why I comment on this particular line.)

Comment (Rossmaxx) [reply ports/libgig/portfile.cmake]: done

File: ports/libgig/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+        ${FEATURE_OPTIONS}
+        -DLIBGIG_BUILD_SHARED=${LIBGIG_BUILD_SHARED}
+)
+
+vcpkg_cmake_install()
+if("extra-tools" IN_LIST FEATURES)
+    vcpkg_copy_tools(TOOL_NAMES dlsdump gigdump gigmerge korg2gig korgdump rifftree sf2dump
```
Comment (Rossmaxx) [inline ports/libgig/portfile.cmake:33]: @dg0yt is this what you mean here

Comment (dg0yt) [reply ports/libgig/portfile.cmake:33]: No. I saw this line, and it is one of the reson why I ask.
Is this for extra tools, or are these all?

Comment (Rossmaxx) [reply ports/libgig/portfile.cmake:33]: I don't know much about what's in here, I just have the pr. Maybe @PhysSong might know more.

Comment (dg0yt) [reply ports/libgig/portfile.cmake:33]: Maybe it is really as simple as renaming the feature to "tools" ?

Comment (Rossmaxx) [reply ports/libgig/portfile.cmake:33]: yeah ig.

File: ports/libgig/vcpkg.json (hunk 2 above)
Comment (Rossmaxx) [reply ports/libgig/vcpkg.json]: I removed the extra and changed it to `tools`

File: ports/libgig/portfile.cmake (hunk 4 above)
Comment (Rossmaxx) [reply ports/libgig/portfile.cmake:33]: done

**Comment by Rossmaxx** [comment]
@dg0yt Changed `extra-toos` to just `tools`. Did i miss anything? Anything else to do before it gets merged?

**Review by LilyWangLL — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks for the new port!

**Pull request merged**